File: backend/app/services/event_service.py
from app.utils.exceptions import ValidationError
from ..models.event_model import Event
from .. import db

def get_all_events(page=1, limit=10):
    query = Event.query.order_by(Event.start_date)
    total = query.count()
    events = query.offset((page - 1) * limit).limit(limit).all()
    return events, total

def create_event(data):
    start_date = data.get("start_date")
    end_date = data.get("end_date")
    if start_date and end_date and end_date < start_date:
        raise ValidationError("End date must be after start date.") 

    event = Event(**data)
    db.session.add(event)
    db.session.commit()
    return event

def update_event(event_id, data):
    start_date = data.get("start_date")
    end_date = data.get("end_date")
    if start_date and end_date and end_date < start_date:
        raise ValidationError("End date must be after start date.") 
    event = Event.query.get_or_404(event_id)
    for key, value in data.items():
        setattr(event, key, value)
    db.session.commit()
    return event

def delete_event(event_id):
    event = Event.query.get_or_404(event_id)
    db.session.delete(event)
    db.session.commit()

def search_events(title, page=1, limit=10):
    query = Event.query.filter(Event.title.ilike(f"%{title}%")).order_by(Event.start_date)
    total = query.count()
    events = query.offset((page - 1) * limit).limit(limit).all()
    return events, total

File: frontend/src/services/eventService.ts
import { eventApi } from "./eventApi";
import { Event } from "../types/Event";

export const eventService = {
  async fetchEvents(page = 1, limit = 10) {
    const res = await eventApi.getAll(page, limit);
    if (!res.success) throw new Error(res.error!.message);
    return res;
  },

  async createEvent(data: Omit<Event, "id">) {
    const res = await eventApi.create(data);
    if (!res.success) throw new Error(res.error!.message);
    return res;
  },

  async updateEvent(id: number, data: Omit<Event, "id">) {
    const res = await eventApi.update(id, data);
    if (!res.success) throw new Error(res.error!.message);
    return res;
  },

  async deleteEvent(id: number) {
    const res = await eventApi.delete(id);
    if (!res.success) throw new Error(res.error!.message);
    return res;
  },

  async searchEvents(query: string, page = 1, limit = 10) {
    const res = await eventApi.search(query, page, limit);
    if (!res.success) throw new Error(res.error!.message);
    return res;
  },
};
